---
title: "Funtional Genomics Part 1: Differential Expression"
output:
  html_document:
    toc: true
    toc_depth: 3
---

# Running Differential Expression Analyses using RNA-seq data

### Precuser

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

## Installing R packages - this has already been done



```{r eval=FALSE}
# if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# BiocManager::install("DESeq2")

```

## Importing the R Libraries

```{r}
library(here)
library(tidyverse) # tidyverse contrains many things - including ggplot2
library(magrittr)

library(DESeq2) # from BiocManager


```

## DESeq2 analyses

```{r}
#DESeq2 analyses#DESeq2 analyses
labonte_meta = read_csv(file = here('data','labonte_dlpfc_meta.csv'))
colnames(labonte_meta)
```

Select the columns we want from the labonte meta-data 

```{r}
labonte_meta %<>% 
  select(geo_accession, 
         expr_names, 
         age = age.ch1, 
         gender = gender.ch1, 
         pmi = pmi.ch1, 
         rin = rin.ch1, 
         ph = ph.ch1, 
         phenotype = phenotype.ch1)

```

#### count data is needed for DEseq2, we obtained this from biojupies

```{r}
labonte_expr = read_tsv( 
  here('data',"GSE102556","GSE102556-expression.counts.ARCHS4.txt.zip"))

labonte_expr %<>% select(c("gene_symbol", labonte_meta$geo_accession))
```

#### ensure meta-data in correct order

```{r}
#ensure right order
labonte_meta %<>% 
  mutate(geo_accession = factor(geo_accession, 
                                levels = colnames(labonte_expr %>%
  select(-gene_symbol))))

labonte_meta %<>% 
  arrange(geo_accession)
```

#### Let's check the order

```{r}
#quick check for order
sum(labonte_meta$geo_accession == colnames(labonte_expr %>% select(-gene_symbol)))
```

#### convert from tibble to data.frame by adding row names

```{r}
#convert from tibble to data.frame by adding row names
labonte_expr %<>% column_to_rownames("gene_symbol") %>% as.data.frame()
```

## Get DESeq DataSet

Using `~ gender + phenotype + ph + rin + pmi + age`

```{r}
dds <- DESeqDataSetFromMatrix(countData = labonte_expr,
                              colData = labonte_meta,
                              design= ~ gender + phenotype + ph + rin + pmi + age)
```

Using `~ gender + phenotype + rin + age` (without ph)

```{r}
dds <- DESeqDataSetFromMatrix(countData = labonte_expr,
                              colData = labonte_meta,
                              design= ~ gender + phenotype +  rin + age)
```

### prefilter data

```{r}
#prefilter data
keep <- rowSums(counts(dds)) >= 50
print(paste("Keeping", sum(keep), "of", nrow(labonte_expr), "genes"))

dds <- dds[keep,]
```

## compute differential expression - may be slow

```{r}
dds <- DESeq(dds)
```

### lists the coefficients / contrasts

```{r}
resultsNames(dds)
```

### get the result for case/control status

```{r}
#get the result for case/control status
res <- results(dds, name="phenotype_MDD_vs_CTRL")
res %<>% as_tibble(rownames = "gene_symbol")

#print results ordered by p-value
res %>% arrange(pvalue)
```

### get the result age

```{r}
#get the result for age
res <- results(dds, name="age")
res %<>% as_tibble(rownames = "gene_symbol")

#print results ordered by p-value
res %>% arrange(pvalue)
```

### examine a histogram of the p-values

```{r}
#examine p-value histogram
hist(res$pvalue)
```

### examining the top gene


```{r}
# pivot_longer (the used to be called "melting") 
melted_expression <- labonte_expr %>% 
  as_tibble(rownames = "gene_symbol") %>% 
  pivot_longer(-gene_symbol, 
               names_to = "geo_accession", 
               values_to = "count") 
```

```{r}
# melted_expression2 <- labonte_expr %>% as_tibble(rownames = "gene_symbol") %>% melt(value.name = "count") %>% as_tibble()
# melted_expression %<>% dplyr::rename(geo_accession = variable)
```

### select our gene of interest

```{r}
expression_RPL23AP4 <- melted_expression %>%
  filter(gene_symbol == "RPL23AP4")
```

### merge the metadata for this gene

```{r}
#merge metadata
expression_RPL23AP4 <- inner_join(expression_RPL23AP4, labonte_meta)
expression_RPL23AP4 %>% arrange(-count) #lots of zeroes
```

```{r}
expression_RPL23AP4 %>% 
  group_by(phenotype) %>% 
  summarize(mean_count = mean(count), 
            max_count = max(count))
```

## Optional Exercise

Think of and apply more stringent prefiltering

---------------------------------------------------------------

## Test polygenic score associations

correlate risk score with expression of each gene

#### Read in the day one results and merge with our gene expression data

```{r}
colnames(labonte_expr)
```


```{r}
## reading in the risk scores from day 1 of the Summer School
risk_scores <- read_csv(here("data", "day1_PRS", "risk_scores.csv"))

## joining the risk scores to the donor meta data
labonte_meta_plus_risk <- inner_join(labonte_meta, risk_scores, 
                                     by = "geo_accession")

## because this could have re-ordered the columns again - we are repeating the steps from above to force the column order back..
labonte_meta_plus_risk %<>% 
  mutate(geo_accession = factor(geo_accession, 
                                levels = colnames(labonte_expr))) %>%
  arrange(geo_accession)
  
```

### add in a term for day one PRS to the design

```{r}
dds_PRS <- DESeqDataSetFromMatrix(countData = labonte_expr,
                              colData = labonte_meta_plus_risk,
                              design= ~ gender + phenotype + ph + rin + pmi + age + day1_PRS)
```


### prefilter as before

```{r}
#prefilter data
keep <- rowSums(counts(dds_PRS)) >= 50
print(paste("Keeping", sum(keep), "of", nrow(labonte_expr), "genes"))

dds_PRS <- dds_PRS[keep,]

dds_PRS_res <- DESeq(dds_PRS)
res <- results(dds_PRS_res, name="day1_PRS")
res %<>% as_tibble(rownames = "gene_symbol")
res %>% arrange(pvalue)
```

### SST is of interest lets look at this gene in depth



